ctapipe/image/muon/processor.py

```diff
@@ -0,0 +1,275 @@
+"""
+High level muon analysis  (MuonProcessor Component)
+"""
+import numpy as np
+
+from ctapipe.containers import (
+    ArrayEventContainer,
+    MuonParametersContainer,
+    MuonTelescopeContainer,
+)
+from ctapipe.coordinates import TelescopeFrame
+from ctapipe.core import QualityQuery, TelescopeComponent
+from ctapipe.core.traits import FloatTelescopeParameter, List, Tuple, Unicode
+
+from .features import (
+    intensity_ratio_inside_ring,
+    mean_squared_error,
+    ring_completeness,
+    ring_containment,
+)
+from .intensity_fitter import MuonIntensityFitter
+from .ring_fitter import MuonRingFitter
+
+INVALID = MuonTelescopeContainer()
+INVALID_PARAMETERS = MuonParametersContainer()
+
+__all__ = ["MuonProcessor"]
+
+
+class ImageParameterQuery(QualityQuery):
+    """
+    For configuring quality checks performed on image parameters
+    before rings are fitted.
+    """
+
+    quality_criteria = List(
+        Tuple(Unicode(), Unicode()),
+        default_value=[
+            ("min_pixels", "dl1_params.morphology.n_pixels > 100"),
+            ("min_intensity", "dl1_params.hillas.intensity > 500"),
+        ],
+        help=(
+            "Quality cuts as list of tuples of ('description', 'expression string'), "
+            "e.g. ``[('mycut', 'x > 3'),]``, to select muon images for analysis. "
+            "You may use ``numpy`` as ``np`` and ``astropy.units`` as ``u``, "
+            "but no other modules. "
+            "DL1 image parameters can be accessed by prefixing the wanted parameter "
+            "with ``dl1_params``. "
+        ),
+    ).tag(config=True)
+
+
+class RingQuery(QualityQuery):
+    """
+    For configuring quality checks performed on the extracted rings before
+    computing efficiency parameters.
+    """
+
+    quality_criteria = List(
+        Tuple(Unicode(), Unicode()),
+        default_value=[
+            ("radius_not_nan", "np.isfinite(ring.radius.value)"),
+            ("min_pixels", "np.count_nonzero(mask) > 50"),
+            ("ring_containment", "parameters.containment > 0.5"),
+        ],
+        help=(
+            "Quality cuts as list of tuples of ('description', 'expression string'), "
+            "e.g. ``[('mycut', 'x > 3'),]``, to select fitted muons for further "
```

Comment:
Sorry for being nitpicky here, but maybe remove the nonsensical example (x won't be defined) and refer to the defaults instead? Or use on the the default lines here.